src/Psalm/Internal/Cache.php

```diff
@@ -42,28 +47,66 @@ public function getItem(string $path)
         }
 
         $cache = Providers::safeFileGetContents($path);
-        if ($this->config->use_gzip) {
+        if ($cache === '') {
+            return null;
+        }
+
+        if ($this->config->compressor === 'deflate') {
             $inflated = @gzinflate($cache);
-            if ($inflated !== false) {
-                $cache = $inflated;
-            }
+        } elseif ($this->config->compressor === 'lz4') {
+            /**
+             * @psalm-suppress UndefinedFunction
+             * @var string|false $inflated
+             */
+            $inflated = lz4_uncompress($cache);
+        } else {
+            $inflated = $cache;
+        }
+
+        // invalid cache data
+        if ($inflated === false) {
+            $this->handleInvalidCache($path);
+
+            return null;
         }
 
         if ($this->config->use_igbinary) {
             /** @var object|false $unserialized */
-            $unserialized = igbinary_unserialize($cache);
+            $unserialized = @igbinary_unserialize($inflated);
         } else {
             /** @var object|false $unserialized */
-            $unserialized = @unserialize($cache);
+            $unserialized = @unserialize($inflated);
+        }
+
+        if ($unserialized === false) {
+            $this->handleInvalidCache($path);
+
+            return null;
         }
 
-        return $unserialized !== false ? $unserialized : null;
+        return $unserialized;
+    }
+
+    protected function handleInvalidCache(string $path): void
+    {
+        $this->deleteItem($path);
+
+        $this->error_count++;
+
+        // if 10 previous items were invalid, abort since the cache is invalid and inform the user
+        // we don't report it to the user immediately, since it can happen that a few files get corrupted somehow
+        // however the impact on performance is minimal, therefore we ignore it
+        if ($this->error_count > 10) {
+            throw new RuntimeException(
```

Comment:
I don't agree with any of this, you both say Psalm will become unusable and that you won't notice. So which is it?

Cache is fallthrough / best effort, if it doesn't work, you get the same experience as you do without cache. The assumption is cache can sometimes fail for whatever reason, but fixing it doesn't require human intervention, it tries to fix itself. CI really likes this because it can then run without you needing to manually fix it, which you might not even be able to easily do, depending on your platform.

In your solution you've picked 10 as the magic number of allowed errors, why is 10 special and not 100 or 5? Why would we reopen the fixed issue? 🤔

Reply:
i agree cache should be disposable by default, so another option maybe? :D

im curious to see how compression plays out in github actions where the cache needs to be pulled also

Reply:
if you agree, why not alter your PR? im not a psalm developer, or not aware of the codebase. just sharing my thoughts _here_ ;)

Reply:
If I understood correctly, this PR could make the cache unreadable (because of differences in caching storage) then makes Psalm crash because said cache is unreadable

This is a much bigger hindrance than just having cache not be used when the cache happens to be corrupted by random chance.

I'm fine with trying to read the cache multiple time if there are reasons to think it could fix itself, but crashing and asking for people (who are probably using CI) to delete the cache folder is not reasonable. 

I'm fine with adding a warning that the cache couldn't be read X times but then Psalm should go on without cache (or possibly delete the cache itself so it doesn't repeat on the next run)

Reply:
Yeah but changing compressor, especially if you change the default one will make the cache unreadable, no? So combined with the second change will make Psalm crash

Reply:
@orklah correct, this was mentioned in the review.

Reply:
maybe this PR should focus on "make compressor configurable" instead, than we improve caching in a separate PR

Reply:
Yeah, let's merge the compressor part alone and then open a discussion about the cache error part. We can see that me merging a PR that did two things in #9889 results in debates, I don't want to make the same mistake again.

@kkmuffme can you revert the cache error part of this PR so we can have a discussion only on the error part? No need to merge something that could be reverted by the next PR after the discussion

Reply:
> ignore errors when reading cache to fix https://github.com/vimeo/psalm/issues/8679, I'm totally fine with that.
> do it consistently then for all places that use/throw on errors in cached data

cant we do it in this PR? or why  insist this PR should revert it first? do we agree cache should be disposable by default already?

>  at least include a message in the output that invalid cache data was encountered but self-healed (for reads) or an error if writes failed, so people are at least aware of it

yes i suggested it could be written to stdout (regardless of --debug flag sounds reasonable)